### ISSUE TITLE
Slow `push` performance on missing target

### ISSUE THREAD

**Issue by mih** [datalad/datalad#4466]
I am not sure if/how it can be done better, but I am recording the fact that I complained first ;-)

It knows immediately that "auto-detection" is the only way to deal with the lack of `--to`, and it also seems to know that there is no tracked remote. It still performs the internal `diff` (which is slow) and only then errors out.

```
% datalad push --since '' --force gitpush 
[INFO   ] No tracked remote for active branch, detection of last pushed state not in effect. 

<quite some time passes>

[WARNING] No push target given, and none could be auto-detected, please specific via --to [publish(/ds/hcp/tmp/hcp_structural_preprocessed)] 
publish(impossible): . (dataset) [No push target given, and none could be auto-detected, please specific via --to]
datalad push --since '' --force gitpush  13.64s user 0.93s system 102% cpu 14.220 total
```

**Comment by yarikoptic** [comment]
If there is no -r Afaik --since had no point (unless it is used to figure out which annexed files to push? But that would be difference in behavior Afaik as well, since should rely on annex wanted) and thus diff shouldn't be used at all.

**Comment by mih** [comment]
Was addressed with #4517
